feat: Make it possible to pass any scikit-learn metric string to `report_metrics`

Muhammad-Rebaal opened the pull request:
Closes #1519 .

---

Hi @auguste-probabl , @MarieSacksick !

I've resolved the issue [#1519](https://github.com/probabl-ai/skore/issues/1519) in  the `report_metric` function to scores the report metrics string. Could you please review the PR .

Thank You !

thomass-dev commented:
Thanks for your contribution. Please can you fix the failing tests and add a dedicated use-case in the unittests?

auguste-probabl commented on skore/src/skore/sklearn/_estimator/metrics_accessor.py:
```diff
@@ -223,27 +232,104 @@ def report_metrics(
-                        f"Invalid metric: {metric!r}. Please use a valid metric"
-                        " from the list of supported metrics: "
-                        f"{list(self._SCORE_OR_LOSS_INFO.keys())}"
+                    # Allow all scikit-learn metrics by checking for prefix patterns
+                    is_sklearn_metric = metric.startswith(
+                        (
+                            "neg_",
+                            "accuracy",
+                            "adjusted_",
+                            "average_",
+                            "balanced_",
+                            "brier_",
+                            "completeness_",
+                            "explained_",
+                            "f1_",
+                            "fowlkes_",
+                            "homogeneity_",
+                            "jaccard_",
+                            "matthews_",
+                            "max_error",
+                            "mean_",
+                            "median_",
+                            "mutual_",
+                            "normalized_",
+                            "precision_",
+                            "r2_",
+                            "rand_",
+                            "recall_",
+                            "roc_auc_",
+                            "top_k_",
+                        )
```
Is this list really needed? Can't we just use `get_scorer` and catch the error if the metric does not exist?

Muhammad-Rebaal replied:
Yes , we can do that . I'll implement that just in a while.

Muhammad-Rebaal commented:
> Thanks for your contribution. Please can you fix the failing tests and add a dedicated use-case in the unittests?

Yes sure , I'll fix them and will add a use-case in the unittest.

auguste-probabl commented:
> Just want to know that is there any test written for this function so that I'd modify that to pass the test ?

Tests would be in `tests/unit/sklearn/estimator/test_estimator.py`

Muhammad-Rebaal commented:
Hi @thomass-dev  , @auguste-probabl  !

I've added the necessary test and fixed the failing tests . Could you please review my PR and let me know if there is anything else that needs to be done.

Thank You !

auguste-probabl commented on skore/src/skore/sklearn/_estimator/metrics_accessor.py:
```diff
@@ -70,23 +70,19 @@ def report_metrics(
         """Report a set of metrics for our estimator.
 
-        Parameters
+                Parameters
```
The indent shouldn't change
```suggestion
        Parameters
```

auguste-probabl commented on skore/src/skore/sklearn/_estimator/metrics_accessor.py:
```diff
@@ -70,23 +70,19 @@ def report_metrics(
             New target on which to compute the metric. By default, we use the target
             provided when creating the report.
-
         scoring : list of str, callable, or scorer, default=None
```
These vertical spaces help readability, please put them back

auguste-probabl commented on skore/src/skore/sklearn/_estimator/metrics_accessor.py:
```diff
@@ -143,12 +134,13 @@ def report_metrics(
+        >>> report.metrics.report_metrics(scoring=["neg_log_loss"])
+                    LogisticRegression
+        Metric
+        Log Loss             -0.10...
```
It should rather say something like
```suggestion
        neg_log_loss             -0.10...
```
or maybe
```suggestion
        Negative Log Loss             -0.10...
```

auguste-probabl commented on skore/src/skore/sklearn/_estimator/metrics_accessor.py:
```diff
@@ -143,12 +134,13 @@ def report_metrics(
+        Log Loss             -0.10...
         """
         if data_source == "X_y":
-            # optimization of the hash computation to avoid recomputing it
-            # FIXME: we are still recomputing the hash for all the metrics that we
-            # support in the report because we don't call `_compute_metric_scores`
-            # here. We should fix it.
```
Can you put the comments back, unless your code really demands changing them?

Muhammad-Rebaal replied:
Thanks for reviewing, I'll put them back

auguste-probabl replied:
this still needs to be done @Muhammad-Rebaal

auguste-probabl commented:
Also noticed that your commits aren't signed. see https://docs.skore.probabl.ai/dev/contributing.html#signing-commits

auguste-probabl commented on skore/src/skore/sklearn/_estimator/metrics_accessor.py:
```diff
@@ -134,11 +143,12 @@ def report_metrics(
                     LogisticRegression
         Metric
-        Log Loss             -0.10...
+        Negative Log Loss      -0.10...        (↘︎)
```
Need to add `Favorability` as a column

Muhammad-Rebaal replied:
Doesn't understand your words can you clarify a little more

auguste-probabl replied:
Yep sorry, please look at the previous example (the one just above) and add "Favorability" the same way

auguste-probabl replied:
You should be able to test this using pytest

auguste-probabl commented on skore/tests/unit/sklearn/estimator/test_estimator.py:
```diff
@@ -1341,3 +1341,61 @@ def test_estimator_report_average_return_float(binary_classification_data):
+    assert result.loc["log_loss", classifier.__class__.__name__] > 0
+
+
+def test_estimator_report_brier_score_requires_binary_classification():
+    """Check that the Brier score raises an error when used with multiclass classification.
+
+    Non-regression test for:
+    https://github.com/probabl-ai/skore/issues/1540
+    """
+    estimator = LogisticRegression()
+    X, y = make_classification(n_classes=3, n_clusters_per_class=1, random_state=42)
+    X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)
+
+    report = EstimatorReport(
+        estimator, X_train=X_train, y_train=y_train, X_test=X_test, y_test=y_test
+    )
+
+    # Check that using brier_score with multiclass raises an error
+    with pytest.raises(
+        ValueError, match="Brier score is only defined for binary classification"
+    ):
+        report.metrics.brier_score()
```
This test has nothing to do with the issue being solved; also, shouldn't it raise an `AttributeError`? Please split this out into another PR if you really want to have it merged

auguste-probabl commented:
please sign your commits

Muhammad-Rebaal commented:
> please sign your commits

Yeah , my last commit was signed.

auguste-probabl commented:
> > please sign your commits
> 
> Yeah , my last commit was signed.

Awesome, thanks! However *all* your commits in the PR have to be signed

Muhammad-Rebaal commented:
> > > please sign your commits
> > 
> > Yeah , my last commit was signed.
> 
> Awesome, thanks! However *all* your commits in the PR have to be signed

Yeah , it should be but besides all that learning signing commits is a new thing that I learned while contributing that's why the initial commits weren't signed. Although, I've added Favorability column and remove the test code which fixes the raising error , hopefully open a new issue for that . Is there anything else needed in this , let me know . I'd be happy to help.
Thank you

thomass-dev commented:
Please, **all** your commits need to be signed.
One solution is to rebase your commits and squash all of it, before re-committing using `git commit -S`.

Muhammad-Rebaal commented:
> Please, **all** your commits need to be signed. One solution is to rebase your commits and squash all of it, before re-committing using `git commit -S`.

Thanks @thomass-dev , I'll take a look at that and implement that .

auguste-probabl commented on skore/tests/unit/sklearn/estimator/test_estimator.py:
```diff
@@ -1341,3 +1341,40 @@ def test_estimator_report_average_return_float(binary_classification_data):
+        *load_breast_cancer(return_X_y=True), random_state=0
+    )
+    classifier = LogisticRegression(max_iter=10_000)
+    classifier.fit(X_train, y_train)
```
```suggestion
```
EstimatorReport does the fitting

auguste-probabl commented on skore/tests/unit/sklearn/estimator/test_estimator.py:
```diff
@@ -1341,3 +1341,40 @@ def test_estimator_report_average_return_float(binary_classification_data):
+    )
+
+    # Use scikit-learn's get_scorer to handle neg_log_loss
+    neg_log_loss_scorer = get_scorer("neg_log_loss")
+
+    result = report.metrics.report_metrics(scoring=[neg_log_loss_scorer])
```
```suggestion
    result = report.metrics.report_metrics(scoring=["neg_log_loss"])
```
The whole point of this change is to support passing scorers as strings.

auguste-probabl commented on skore/tests/unit/sklearn/estimator/test_estimator.py:
```diff
@@ -1341,3 +1341,40 @@ def test_estimator_report_average_return_float(binary_classification_data):
+    assert "log_loss" in result.index
+
+    # Check that the value matches the absolute value of neg_log_loss
+    assert result.loc["log_loss", classifier.__class__.__name__] > 0
```
That code doesn't do what the comment says.

Muhammad-Rebaal replied:
Yes , I've fixed that . It might be a mistake or so.

auguste-probabl commented on skore/tests/unit/sklearn/estimator/test_estimator.py:
```diff
@@ -1341,3 +1341,40 @@ def test_estimator_report_average_return_float(binary_classification_data):
+
+
+def test_estimator_report_metric_with_neg_metrics():
+    """Check that scikit-learn metrics with 'neg_' prefix are handled correctly.
+
+    Non-regression test for:
+    https://github.com/probabl-ai/skore/issues/1519
+    """
```
```suggestion
    """Check that scikit-learn metrics with 'neg_' prefix are handled correctly."""
```

auguste-probabl commented on skore/tests/unit/sklearn/estimator/test_estimator.py:
```diff
@@ -1341,3 +1341,40 @@ def test_estimator_report_average_return_float(binary_classification_data):
+    Non-regression test for:
+    https://github.com/probabl-ai/skore/issues/1519
+    """
+    from sklearn.datasets import load_breast_cancer
+    from sklearn.linear_model import LogisticRegression
+    from sklearn.metrics import get_scorer
+    from sklearn.model_selection import train_test_split
```
Move these outside

Muhammad-Rebaal replied:
Yes, I've updated that and also remove redundancy.

auguste-probabl commented on skore/src/skore/sklearn/_estimator/metrics_accessor.py:
```diff
@@ -223,33 +221,79 @@ def report_metrics(
                     if scoring_kwargs is None:
                         metrics_kwargs = {}
                     else:
-                        # check if we should pass any parameters specific to the metric
```
put the comment back please

Muhammad-Rebaal replied:
yeah sure

Muhammad-Rebaal commented:
Hi @auguste-probabl !

Could you please review this PR and let me know if there is anything else needed to resolve ?

Thank You !

auguste-probabl commented on skore/tests/unit/sklearn/estimator/test_estimator.py:
Your PR makes it possible to write `report.metrics.report_metrics(scoring=["some_metric"])` (i.e. a string), so you have to test this

auguste-probabl reviewed (changes requested):
Missing a test and some comments were removed which should not be

auguste-probabl commented on skore/tests/unit/sklearn/estimator/test_estimator.py:
```diff
@@ -1341,3 +1342,97 @@ def test_estimator_report_average_return_float(binary_classification_data):
         assert isinstance(result, float)
+
+
+def test_estimator_report_metric_with_neg_metrics():
+    """Check that scikit-learn metrics with 'neg_' prefix are handled correctly."""
+
+    X_train, X_test, y_train, y_test = train_test_split(
+        *load_breast_cancer(return_X_y=True), random_state=0
+    )
+    classifier = LogisticRegression(max_iter=10_000)
+
+    report = EstimatorReport(
+        classifier,
+        X_train=X_train,
+        y_train=y_train,
+        X_test=X_test,
+        y_test=y_test,
+    )
+
```
You should be able to use the `binary_classification_data` fixture:
```suggestion
def test_estimator_report_metric_with_neg_metrics(binary_classification_data):
    """Check that scikit-learn metrics with 'neg_' prefix are handled correctly."""

    classifier, X_test, y_test = binary_classification_data

    report = EstimatorReport(
        classifier,
        X_test=X_test,
        y_test=y_test,
    )
```
You might need to use a different metric than log loss since the estimator is a RandomForestClassifier.

Same for the other test.

auguste-probabl commented on skore/tests/unit/sklearn/estimator/test_estimator.py:
```diff
@@ -1341,3 +1342,97 @@ def test_estimator_report_average_return_float(binary_classification_data):
+    favorability = result_multi.loc["Log Loss"]["Favorability"]
+    assert favorability == "(↘︎)"
+
+    # Regression test case
+    X_reg, y_reg = make_regression(random_state=42)
+    X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(
+        X_reg, y_reg, random_state=42
+    )
+    regressor = LinearRegression()
+    reg_report = EstimatorReport(
+        regressor,
+        X_train=X_train_reg,
+        y_train=y_train_reg,
+        X_test=X_test_reg,
+        y_test=y_test_reg,
+    )
+
+    # Test regression metrics
+    reg_result = reg_report.metrics.report_metrics(
+        scoring=["neg_mean_squared_error", "neg_mean_absolute_error", "r2"],
+        indicator_favorability=True,
+    )
+
+    assert "Mean Squared Error" in reg_result.index.get_level_values(0)
+    assert "Mean Absolute Error" in reg_result.index.get_level_values(0)
+    assert "R²" in reg_result.index.get_level_values(0)  # Changed from "R2" to "R²"
+
+    # Check favorability
+    assert reg_result.loc["Mean Squared Error"]["Favorability"] == "(↘︎)"
+    assert reg_result.loc["R²"]["Favorability"] == "(↗︎)"  # Changed from "R2" to "R²"
```
Make a new test for this case.

auguste-probabl commented on skore/tests/unit/sklearn/estimator/test_estimator.py:
```diff
@@ -1341,3 +1342,92 @@ def test_estimator_report_average_return_float(binary_classification_data):
+    assert np.isclose(log_loss_value, abs(neg_log_loss_value))
+
+
+def test_estimator_report_with_sklearn_scoring_strings():
+    """Test that scikit-learn metric strings can be passed to report_metrics."""
+    # Classification test case
+    X_class, y_class = load_breast_cancer(return_X_y=True)
+    X_train, X_test, y_train, y_test = train_test_split(
+        X_class, y_class, random_state=42
+    )
```
Same as earlier, use a fixture

auguste-probabl commented on skore/tests/unit/sklearn/estimator/test_estimator.py:
```diff
@@ -1341,3 +1342,92 @@ def test_estimator_report_average_return_float(binary_classification_data):
+    favorability = result_multi.loc["Log Loss"]["Favorability"]
+    assert favorability == "(↘︎)"
+
+
+def test_estimator_report_with_sklearn_scoring_strings_regression():
+    """Test scikit-learn regression metric strings in report_metrics."""
+    X_reg, y_reg = make_regression(random_state=42)
+    X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(
+        X_reg, y_reg, random_state=42
+    )
+    regressor = LinearRegression()
+    reg_report = EstimatorReport(
+        regressor,
+        X_train=X_train_reg,
+        y_train=y_train_reg,
+        X_test=X_test_reg,
+        y_test=y_test_reg,
+    )
```
Use a fixture

auguste-probabl commented on skore/tests/unit/sklearn/estimator/test_estimator.py:
```diff
@@ -1341,3 +1342,115 @@ def test_estimator_report_average_return_float(binary_classification_data):
+    assert np.isclose(log_loss_value, abs(neg_log_loss_value))
+
+
+@pytest.fixture
+def classification_train_test_data():
+    """
+    Create a classification dataset and return fitted estimator,
+    training and test data.
+    """
+    X_class, y_class = load_breast_cancer(return_X_y=True)
+    X_train, X_test, y_train, y_test = train_test_split(
+        X_class, y_class, random_state=42
+    )
+    return (
+        LogisticRegression(max_iter=10_000).fit(X_train, y_train),
+        X_train,
+        X_test,
+        y_train,
+        y_test,
+    )
```
There is already a fixture for binary classification, you just used it in the other test above. Same for regression, there is already a fixture you can use.
